### PR TITLE
Allow AST plugins to manipulate Javascript scope

### PR DESCRIPTION
**Pull request by ef4**
This spikes out the ability for AST plugins to access Javascript imports and expression.

Still needs more practical testing.

Does not support Ember < 3.28 because that's where we gained the option to use lexically scoped values in non-strict templates.

### CODE REVIEW

**Comment by ef4** [comment]
The usefulness of this will depend on getting ember-cli-htmlbars to process all templates through babel-plugin-ember-template-compilation. At the moment:

 - [ ] route templates
 - [ ] non-colocated component templates
 - [x] co-located component templates
 - [x] all inline (in JS) templates

**Comment by ef4** [comment]
(All those cases already work in embroider, but addons can't realistically adopt if we don't make it work in classic builds too)

**Comment by ef4** [comment]
I confirmed in a test app that https://github.com/ember-cli/ember-cli-htmlbars/pull/749 is enough to make this new feature work in all templates including the stand-alone, non-colocated ones.

File: __tests__/tests.ts [hunk 1]
```diff
@@ -585,6 +589,143 @@ describe('htmlbars-inline-precompile', function () {
...
+            },
+          },
+        };
+      });
+
+      let transformed = transform(stripIndent`
+        import { precompileTemplate } from '@ember/template-compilation';
+        export default function() {
+          let two = 'twice';
+          const template = precompileTemplate('<Message @text={{onePlusOne}} />');
```
Comment (dfreeman) [inline __tests__/tests.ts:721]: What about bindings within the template itself?

```hbs
{{#let "two" as |two|}}
  <Message @text={{onePlusOne}} />
{{/let}}
```

Comment (ef4) [reply __tests__/tests.ts:721]: Good idea, we need to avoid the intersection of both the js and hbs existing bindings.

**Comment by ef4** [comment]
ember-cli-babel 6.1 shipped with the changes required to make this useful in all templates.

**Comment by ef4** [comment]
My next action here is to test this new API in Embroider, since we already do the same class of things in a less elegant way.

**Comment by ef4** [comment]
Embroider is now happily refactored onto this feature as the foundational primitive for getting static components, helpers, and modifiers into templates. See https://github.com/embroider-build/embroider/pull/1276

**Pull request merged**